gmail/stream_implementer.bal

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) 2021 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+class MessageStream {
+    private Message[] currentEntries = [];
+    private int index = 0;
+    private final http:Client httpClient;
+    private final string userEmailId;
+    private final MsgSearchFilter? filter;
+    private string? pageToken = ();
+
+
+    isolated function init(http:Client httpClient, string userId, MsgSearchFilter? filter) returns error? {
+        self.httpClient = httpClient;
+        self.filter = filter;
+        if (self.filter?.pageToken is string) {
+            self.pageToken = <string>self.filter?.pageToken;
+        }
+        self.userEmailId = userId;
+        self.currentEntries = check self.fetchMessages();
+    }
+
+    public isolated function next() returns @tainted record {| Message value; |}|error? {
```

Comment:
When would it return an error? When there is no more messages?

Reply:
No. When an error occurred while iteration/fetch.